[JUP-120] Update logging levels

tommycodebox opened the pull request:
## 🧠  Description
- Use simple logger for all logging levels

![image](https://user-images.githubusercontent.com/52791021/135101486-31cc0cdd-9c75-41a1-827a-63de2b269ff6.png)

## 🕵🏻‍♂️  Task
>#### [[JUP-120] Update logging levels](https://darcs.atlassian.net/browse/JUP-120)

arijoon commented on index.js:
```diff
@@ -42,57 +42,57 @@ have been logged anyway.
+const parseArgs = (args) => args
+  .map(arg => stringify(
+    arg instanceof Error ? {
+      name: arg.name,
```
could an `Error` object have more properties than these 3 that were added as context?

arijoon replied:
Actually ignore me there, it seems that is not even a valid use case of `Error`, and we have some of that in our codebase. An extension of `Error` could have more properties however so may want to add a `context` field there that we can populate in our custom errors such as `DVFError`

adrian-gierakowski replied:
There are many error classes in libraries that include extra props (like MongoError) which might be useful to long.

adrian-gierakowski replied:
Btw. I'm not sure about simply concatenating all args into a string is sticking it all into message. I think we should be able to pass in objects and have those logged in structured way.

arijoon commented on package.json:
```diff
@@ -7,6 +7,7 @@
   "license": "MIT",
   "dependencies": {
-    "debug": "^4.1.1"
+    "debug": "^4.1.1",
```
can remove this dep if no longer used

arijoon commented on index.js:
```diff
@@ -41,80 +41,58 @@ enabled. This way we can avoid paying for pre-processing if the value would not
   const {
-    prefix = 'dvf',
     root = cwd,
-    logToStderr = false
```
do we no longer need to print anything to stderr?

tommycodebox replied:
I dont think we need, if we dump json with correct fields and GCP parses that

tommycodebox replied:
and I checked codebase where this package was required and there were no cases when options were passed

adrian-gierakowski commented on index.js:
```diff
@@ -52,73 +52,49 @@ const root =
+  const logger = (...args) => {
+    const message = parseArgs(args)
+    const payload = { severity, timestamp: Date.now(), context, message }
+    console.log(stringify(payload))
```
I'd like to be able to toggle (with an env var) between logging as single line of JSON and pretty printing. This is really useful when running things locally.

tommycodebox replied:
so now by adding `PRETTY=true` you will get json indented with 2 places

tommycodebox commented on package.json:
```diff
@@ -7,6 +7,6 @@
   "license": "MIT",
   "dependencies": {
-    "debug": "^4.1.1"
+    "safe-json-stringify": "^1.2.0"
```
I removed that `.enabled` at all, as we dont seem to use it, and I created simple pattern to account for `DEBUG` env var to allow logging all levels with `DEBUG=*` and specific levels in isolation e.g. `DEBUG=WARN`

by default we would log following levels (basically all without DEBUG level)
`[ LOG, WARN, ERROR, EMERGENCY ]`

adrian-gierakowski replied:
I'm sorry, I'm not going to approve this PR if we don't get feature parity with existing logger. It's very simple the achieve this: bring back the debug module and use it's `.enabled` prop to check if you should log or not (even though you don's use it for actual logging).

adrian-gierakowski replied:
btw. if we use `debug` module (and/or `DEBUG` env var) we'll need the `dvf:` prefix as there are many other modules in the JS ecosystem using `debug` and if we don't prefix we might accidentally enable logging in one of the libraries we used. Keeping it backward compatible will also save you time since you will not need to update env vars in existing services which use `dvf-logger`

arijoon commented on index.js:
```diff
@@ -52,79 +52,73 @@ const root =
+      (LEVELS.includes(DEBUG) && DEBUG === severity)
+    ) {
+      const log_args = [payload]
+      if (PRETTY) log_args.push(null, 2)
```
Can we use a different logger format in case of `PRETTY`? 
```
timestamp | context: message | JSON.stringify(args)
```
instead of a json dump of everything to make it much more readable locally since it won't be auto parsed

adrian-gierakowski commented on index.js:
```diff
@@ -52,79 +52,79 @@ const root =
-    // no-op
-    () => {}
-}
+const DEBUG = process.env.DEBUG
```
after these changes, the lazy loggers don't do what they were designed to do: `expandThunks` should have only been called if logger was enabled

adrian-gierakowski reviewed (changes requested):
the following comment has not been sufficiently addressed:
https://github.com/DeversiFi/logger/pull/2/files#r734354894

adrian-gierakowski commented on index.js:
```diff
@@ -52,79 +52,79 @@ const root =
+
+const makeStrictLogger = (severity, context) => {
+  const logger = (...args) => {
+    const message = parseArgs(args)
+    const payload = { severity, timestamp: Date.now(), context, message }
+
+    if (
+      (!DEBUG && severity !== 'DEBUG') ||
+      (DEBUG && !LEVELS.includes(DEBUG) && severity !== 'DEBUG') ||
+      DEBUG === '*' ||
+      (LEVELS.includes(DEBUG) && DEBUG === severity)
+    ) {
```
llthis is not efficient: this loggic is executed every single time the logging function is called, even though eventually we might end up not logging anything. Original logger implementation decides if particular logger instance is `enabled` or not at the instantiation time (once per process run) and then simply checks `.enabled` prop to decide if logging logic should be executed or not

tommycodebox commented:
@adrian-gierakowski I have addressed your feedback by optimizing the logger using `.enabled` but without using `debug`, logic now also wont log in test environment

arijoon commented:
> @adrian-gierakowski I have addressed your feedback by optimizing the logger using `.enabled` but without using `debug`, logic now also wont log in test environment

I don't think that will bring back the feature @adrian-gierakowski was asking for which is pattern matching the files to log based on their name, e.g. `DEBUG='dvf:ERROR:*:serviceX.js'`

tommycodebox commented:
ok, if we are using this feature, lets then use `.enabled` of that `debug` package to determine whether to log or not 👍🏻

arijoon commented:
Flattening of the arguments object: https://darcs.slack.com/archives/GL7JLS5RD/p1637682852493000

arijoon commented:
Is this ready for merge?

tommycodebox commented:
@arijoon this is finished from my side but was still pending review from you and @adrian-gierakowski so we are clear on this

arijoon commented on index.js:
```diff
@@ -52,79 +53,113 @@ const root =
-  elem => typeof elem == 'function' ? elem() : elem
-)
+const PRETTY = Boolean(process.env.PRETTY)
+const DEPTH = process.env.DEPTH
```
does this need to be defaulted?

tommycodebox replied:
nope, its defaulted in `stringify` function

arijoon commented on stringify.js:
```diff
@@ -0,0 +1,24 @@
+const stringify = (data, depth = 5) => {
+  return data === undefined
+    ? 'null'
```
`'undefined'` would be more appropriate

tommycodebox replied:
JSON doesnt support `undefined`, but does support `null` and Im not sure we need a string of `'undefined'` 🤔

arijoon reviewed (approved):
Looks great to me @tommycodebox , lets put it in stark trading `dev` and see how it looks in action

arijoon commented on yarn.lock:
```diff
@@ -2,14 +2,19 @@
   resolved "https://registry.yarnpkg.com/ms/-/ms-2.1.2.tgz#d09d1f357b443f493382a8eb3ccd183872ae6009"
   integrity sha512-sGkPx+VjMtmA6MX27oA4FBFELFCZZ4S4XqeGOXCv68tT+jb3vk/RyaKWP0PTKyWtmLSM0b+adUTEvbs1PEaH2w==
+
+safe-json-stringify@^1.2.0:
```
btw this doesn't seem to be used anymore so can it be removed?
